Rework erinyes into avenging Furies

entrez opened the pull request:
Based on conversations on IRC with paxed, bhaak, K2, and Umbire.  This is an idea based on the role of the erinyes in Greek myth: punishers of oathbreakers, hosts who harm their guests, murderers, and others who have violated important ancient social mores or customs.  This is reflected in NetHack by the erinyes becoming more dangerous as you abuse your alignment -- something like killing a peaceful monster does seem consistent to me with things that bring down the Furies in mythology.  The alignment abuse metric that determines the difficulty of the erinyes is tracked separately from alignment record, and only counts negative adjustments to alignment: 'adjalign(-5); adjalign(10); adjalign(-5);' would leave you with an alignment record of 0, but alignment abuse value of 10.

If extinction allows, one or more erinyes are also (edit: potentially, based on a roll against abuse value) summoned when you don a helm of opposite alignment (since you're breaking your oath to your god, so it especially incites them).  The helm of opposite alignment also causes brief confusion, and #offering is made impossible while confused, in order to make this a potential challenge on Astral rather than something which can easily be blown past by #offering the amulet on the next turn.

This could use more testing and conversation about possible changes or extensions (or alternatives), but I think it's at a point now where I am comfortable submitting the patch in its current state and starting some discussion around it.

entrez committed:
Make HoOA confuse the hero, summon erinyes

A helmet psychically reaching inside your head and twiddling knobs in
your brain must be a confusing experience.  And the instant rejection of
your god (not to mention the vow that you made to find the amulet for
them), perhaps on the very cusp of their ascendance over the other gods,
is sort of the ultimate oathbreaking, so it interests the erinyes.

entrez committed:
Require a clear head to #offer

This will increase the potential challenge involved in donning a helm of
opposite alignment on Astral, now that it confuses the hero -- it means
that you must spend at least a turn after putting on the helm clearing
the confusion.  Since putting on the helm of opposite alignment may also
summon erinyes, this hopefully turns it from a situation where it's a
no-brainer to immediately ditch them by #offering the Amulet of Yendor
into one where the player must actually stop and think a little bit.

entrez committed:
Make erinyes scale with alignment abuse

Consistent with their mythological role of punishing those who had
violated societal taboos -- oathbreakers, hosts who attacked their
guests, etc -- erinyes scale with the cumulative amount of alignment
abuse the hero has committed over the course of the game.  This is
tracked separately from the alignment record, and cannot be cleared by
the hero improving her favor with her god via "good deeds" as the normal
alignment record can.  Erinyes will gain abilities, levels, and attacks
as the hero's alignment abuse worsens.  They will also aggravate
monsters when near the hero.

entrez committed:
Teleport-control monsters resist being teleported

This makes sense to me, and further increases the challenge of erinyes
being summoned on astral after donning a helm of opposite alignment (if
the player has abused his alignment to the extent that erinyes have
teleport control), since they wouldn't be able to be banished across the
map simply by breaking a wand of teleport control.

entrez committed:
Add new erinys encyclopedia entry

entrez commented:
This would break saves.  I have another PR open that breaks saves which
increments EDITLEVEL in the PR itself, #1140, but I didn't change it
this time because I didn't want to be continually changing it if some
other EDITLEVEL bumps come down the pipe while the PR is open.

pat-rankin commented:
Doesn't this let someone summon an erinys as soon as they find a helm of opposite alignment?  And assuming they survive, repeat two more times.

Survival is far from guaranteed though.  It also turns a very early helm of opposite alignment that the hero tries on become almost certain death.

nhmall commented:
This is not a comment on the pull request itself, just my opinion about what would be optimal procedure.
> I have another PR open that breaks saves which
> increments EDITLEVEL in the PR itself, #1140, but I didn't change it
> this time because I didn't want to be continually changing it if some
> other EDITLEVEL bumps come down the pipe while the PR is open.

I think taking steps to draw attention to the requirement to bump EDITLEVEL in include/patchlevel.h is the smoothest way to go, and leave out include/patchlevel.h from any pull requests. It has a high probability of being outdated by the time the pull request is merged.

I've also added a **_++EDITLEVEL when merging_** label that can be applied, although I cannot guarantee it will be applied in every case, or applied by all developers.

entrez commented:
@pat-rankin Yes, it would allow the hero to summon an erinys at will (which could be a strategic decision to avoid their summoning on Astral).  Extremely early-game erinyes may not be a huge problem because they are generated peaceful if you haven't yet committed any alignment abuse, but it could definitely be dangerous.  The abuse-based scaling could be increased further so that their base level, speed, etc, pre-abuse is additionally reduced, so that it's possible to flee from them if you don a HoOA too early.

The erinyes summoning could also be restricted entirely to Astral, of course -- I thought it could be an interesting decision though, whether you want to deliberately don an HoOA repeatedly to clear out the erinyes in preparation for using one on Astral, but also lose all your protection etc.  If donning an HoOA significantly abused alignment it could be even riskier and more interesting, because each time you put it on you'd make the erinyes summons left to go more dangerous!  Maybe some would opt to only eliminate one of the three, etc.  I didn't add this out of concern for its effects on early-game accidental HoOA use -- but it could be something that only abuses your alignment after the erinys is summoned, so only affects subsequent erinys generation, and only increments 'abuse' without decrementing 'record' (since it wouldn't make much sense that wearing the helm harms your alignment record with the god you've just converted to).  Basically just spitballing there.

entrez commented:
I just force-pushed to restore a comment in uchangealign(attrib.c) that
I had apparently deleted by mistake.  I didn't make any other changes.

entrez commented:
Pat, to your point about early-game HoOA use, maybe it could be changed so putting on a helm of opposite alignment just has a (alignment abuse-based?) chance of summoning an erinys, if you're not on Astral.  That would make it more challenging to eliminate all erinyes and make it less "guaranteed insta-death" for the early game.  It could also work with the HoOA alignment abuse idea I mentioned in my previous comment (i.e. it would mean subsequent uses of a helm would have increased likelihood of summoning an erinyes each time if the chance is based on abuse, but also make them more dangerous).

entrez committed:
Add randomness to HoOA erinyes summoning

When not on Astral, make it a roll based on alignment abuse whether an
erinys is summoned.  Also make helm of opposite alignment use contribute
to alignment abuse, so that repeated uses to attempt to clear erinyes
before astral will make them stronger.

entrez committed:
Eliminate magic numbers from uchangealign()

NullCGT commented:
This is a really interesting and flavorful change. Is there any way to communicate the changes in Erinyes strength to the player? It seems like it could be fairly unintuitive for unspoiled players to have Erinyes be much stronger or weaker with no indication of why.

k21971 commented:
> Is there any way to communicate the changes in Erinyes strength to the player? It seems like it could be fairly unintuitive for unspoiled players to have Erinyes be much stronger or weaker with no indication of why.

I don't believe that would be necessary. Players will learn via interaction. There's no in-game indication that elementals are stronger on their home plane; players figured this out as they played. I think the same logic would apply here.

entrez commented:
I'd figure that if you've played enough NetHack to be surprised at the variation in the difficulty, a savvy encyclopedia-using player could likely come up with a hypothesis that their "wrath" is somehow based on how many "deeds unjust and base" the hero has committed, based on the encyclopedia entry in the PR.  If you're at that point as an unspoiled player you are probably very familiar with the encyclopedia and how to use it to make guesses about the game, since I think significant variation in difficulty that is clearly not tied to xlvl would require reaching the lategame across multiple different games.  And I think puzzling out the quote encyclopedia entries and how they might inform you about what you're looking at is a really fun part of being unspoiled about some aspect of the game.

nhmall commented:
A clever true rumor can't hurt though, if you can come up with one.

entrez commented:
Oh yeah, that's a great idea.  I'll try to think of something.  I guess I really only think about the encyclopedia since my recent "unspoiled" experience is with changes in variants, and trying to get relevant true rumors there is tough because they're massively outnumbered by vanilla rumors, even if the variant dev keeps the rumors files updated.  So I basically rely on the encyclopedia.

entrez committed:
Add true rumor about erinyes and alignment abuse

I don't think this entirely gives the game away, but strongly suggests a
connection between bad deeds and how dangerous erinyes are (especially
combined with the new erinys encyclopedia entry).

entrez commented:
Last force-push there was just to resolve merge conflicts from the recent extern.h changes.  (The previous force-push was to remove some formatting fixes that were already merged in 275713b, and didn't introduce any material changes to the patch either.)

paxed commented:
the HoOA thing doesn't really feel right to me, not sure ... I'd also want to lean towards the shadow and darkness aspects of Erinyes from mythology

entrez commented:
What are the shadow and darkness aspects of the Furies in mythology?  I'm not familiar with that.

paxed commented:
> What are the shadow and darkness aspects of the Furies in mythology? I'm not familiar with that.

https://en.wikipedia.org/wiki/Erinyes#Description

entrez commented:
Interesting.  I'm sure that could be cool -- right now there is basically nothing about them other than their number, so anything at all would be a flavor improvement.  There's also no reason these two concepts would have to be mutually exclusive.

The reason I think the HoOA change could be good is that it introduces an additional thing to deal with on Astral if going for an altar other than your own, one that can be tackled in a number of different ways -- if you are planning to use a HoOA on Astral, do you want to risk making the erinyes more dangerous for the chance to potentially eliminate some of them early, do you instead want to try to be careful to avoid alignment abuse so that they aren't particularly strong when you fight them (and/or are weak enough that it will be trivial to use a unicorn horn and #offer the amulet without fighting them), or do you want to spend a significant amount of time trying to generate them normally without buffing them the way HoOA use would?  Do you have a sense of your alignment abuse so far, to know which options are viable?  If you end up on the wrong altar on Astral and have a wish, do you want to risk fighting them or continue to the other altars?  Those are interesting decisions that can add layers of short- and long-term strategy to the end game, which often feels lacking in strategic choices like that by the time you've assembled an ascension kit.  And I think it does make flavor sense that HoOA use can trigger that kind of divine vengeance, given the way conversion is already treated as an especially serious betrayal of your deity in the game.  But it's not necessarily an integral part of the patch (though the reason I wrote all this out is in its defense is because I think it's would add dynamism and interesting decisions to make).

entrez commented:
I will say, though, I think this is a significant enough change and with enough of a delicate sort of balance that it would probably be better implemented in a variant and tested that way first rather than sticking it directly into NetHack "raw".  I think K2 was saying he may implement it in EvilHack -- that's not very equivalent to vanilla in terms of power levels of the hero and monsters, so probably it would settle on different numbers, but it could be interesting to see how well the general concept works there.
